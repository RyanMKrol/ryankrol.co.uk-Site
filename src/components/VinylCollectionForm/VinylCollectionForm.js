/* eslint-disable no-alert */

import React from 'react';
import { Formik, Form } from 'formik';
import fetch from 'node-fetch';

import { baseEndpoint } from '../../utils/ProdUtils';
import { RatingsFormItem, PASSWORD_TYPE } from '../RatingsFormItem';

import styles from './VinylCollectionForm.module.css';

const VINYL_API_ENDPOINT = `${baseEndpoint()}/vinyl`;

const VINYL_INPUT_CONFIG = [
  { itemName: 'title', itemTitle: 'Record Title', itemType: 'text' },
  { itemName: 'artist', itemTitle: 'Artist', itemType: 'text' },
  { itemName: 'mood', itemTitle: 'Mood', itemType: 'text' },
  { itemName: 'password', itemTitle: 'Password', itemType: PASSWORD_TYPE },
];

/**
 * Generates the initial form values
 *
 * @returns {object} The default value for each form item
 */
function generateInitialFormValues() {
  return VINYL_INPUT_CONFIG.reduce((acc, item) => {
    acc[item.itemName] = '';
    return acc;
  }, {});
}

/**
 * Validates the current form values
 *
 * @param {Array<object>} values An array of values currently in the form
 * @returns {object} An array of form item names that need validation
 */
export function validateFormValues(values) {
  return VINYL_INPUT_CONFIG.reduce((acc, item) => {
    if (typeof values[item.itemName] === 'undefined') {
      acc[item.itemName] = 'Required';
    }
    return acc;
  }, {});
}

export const onSubmitHandler = (values, { setSubmitting }) => fetch(VINYL_API_ENDPOINT, {
  method: 'post',
  body: JSON.stringify(values),
  headers: { 'Content-Type': 'application/json' },
})
  .then((res) => res.json())
  .then((json) => window.alert(json.message))
  .then(() => setSubmitting(false));

/**
 * Generates the form fields
 *
 * @param {Array<any>} values Current values in the form
 * @returns {string} JSX for an individual form field
 */
function generateFormFields(values) {
  return VINYL_INPUT_CONFIG.map((item) => (
    <RatingsFormItem
      fieldName={item.itemName}
      fieldTitle={item.itemTitle}
      fieldType={item.itemType}
      currentValue={values[item.itemName]}
    />
  ));
}

/**
 * UI Component representing a form to input vinyl items
 *
 * @returns {string} The UI component
 */
function VinylCollectionForm() {
  return (
    <Formik
      initialValues={generateInitialFormValues()}
      validate={validateFormValues}
      onSubmit={onSubmitHandler}
    >
      {({ isSubmitting, values }) => (
        <Form className={styles.form}>
          {generateFormFields(values)}
          <div className={styles.formItem}>
            <button type="submit" disabled={isSubmitting}>
              Submit
            </button>
          </div>
        </Form>
      )}
    </Formik>
  );
}

export default VinylCollectionForm;
