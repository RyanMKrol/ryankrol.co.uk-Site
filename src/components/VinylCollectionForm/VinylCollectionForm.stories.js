import VinylCollectionForm from './VinylCollectionForm';

export default {
  title: 'Components/VinylCollectionForm',
};

export const Default = () => <VinylCollectionForm />;
