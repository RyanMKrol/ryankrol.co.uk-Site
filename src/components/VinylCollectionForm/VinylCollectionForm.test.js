import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Formik, Form } from 'formik';
import fetch from 'node-fetch';

import VinylCollectionForm, { validateFormValues, onSubmitHandler } from './VinylCollectionForm';

jest.mock('node-fetch');

it('renders VinylCollectionForm', () => {
  const listensItem = mount(
    <Formik>
      <Form>
        <VinylCollectionForm />
      </Form>
    </Formik>,
  );
  expect(toJson(listensItem)).toMatchSnapshot();
});

it('validates values correctly', () => {
  const testItems = {
    title: 'test-title',
    artist: 'test-artist',
    mood: 'test-mood',
    password: undefined,
  };
  const result = validateFormValues(testItems);

  expect(result).toEqual({
    password: 'Required',
  });
});

it('fetches data correctly', async () => {
  const fetchJsonMock = {
    message: 'test-message',
  };
  const windowAlertMock = jest.fn();
  const setSubmittingMock = jest.fn();

  fetch.mockResolvedValueOnce({ json: () => fetchJsonMock });
  jest.spyOn(window, 'alert').mockImplementation(windowAlertMock);

  const setSubmittingBlob = {
    setSubmitting: setSubmittingMock,
  };

  const values = [1, 2, 3];
  await onSubmitHandler(values, setSubmittingBlob);

  expect(windowAlertMock).toBeCalledWith('test-message');
  expect(setSubmittingMock).toBeCalled();
});
