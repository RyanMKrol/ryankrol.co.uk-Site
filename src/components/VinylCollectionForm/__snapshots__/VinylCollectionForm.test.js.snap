// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders VinylCollectionForm 1`] = `
<Formik>
  <Form>
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <VinylCollectionForm>
        <Formik
          initialValues={
            Object {
              "artist": "",
              "mood": "",
              "password": "",
              "title": "",
            }
          }
          onSubmit={[Function]}
          validate={[Function]}
        >
          <Form
            className="form"
          >
            <form
              action="#"
              className="form"
              onReset={[Function]}
              onSubmit={[Function]}
            >
              <RatingsFormItem
                currentValue=""
                fieldName="title"
                fieldTitle="Record Title"
                fieldType="text"
              >
                <div
                  className="formItem highlighted"
                  key="title"
                >
                  Record Title
                  <br />
                  <Field
                    autoComplete="off"
                    name="title"
                    type="text"
                  >
                    <input
                      autoComplete="off"
                      name="title"
                      onBlur={[Function]}
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </Field>
                  <FormikConnect(ErrorMessageImpl)
                    component={[Function]}
                    name="title"
                  >
                    <ErrorMessageImpl
                      component={[Function]}
                      formik={
                        Object {
                          "dirty": false,
                          "errors": Object {},
                          "getFieldHelpers": [Function],
                          "getFieldMeta": [Function],
                          "getFieldProps": [Function],
                          "handleBlur": [Function],
                          "handleChange": [Function],
                          "handleReset": [Function],
                          "handleSubmit": [Function],
                          "initialErrors": Object {},
                          "initialStatus": undefined,
                          "initialTouched": Object {},
                          "initialValues": Object {
                            "artist": "",
                            "mood": "",
                            "password": "",
                            "title": "",
                          },
                          "isSubmitting": false,
                          "isValid": true,
                          "isValidating": false,
                          "registerField": [Function],
                          "resetForm": [Function],
                          "setErrors": [Function],
                          "setFieldError": [Function],
                          "setFieldTouched": [Function],
                          "setFieldValue": [Function],
                          "setFormikState": [Function],
                          "setStatus": [Function],
                          "setSubmitting": [Function],
                          "setTouched": [Function],
                          "setValues": [Function],
                          "status": undefined,
                          "submitCount": 0,
                          "submitForm": [Function],
                          "touched": Object {},
                          "unregisterField": [Function],
                          "validateField": [Function],
                          "validateForm": [Function],
                          "validateOnBlur": true,
                          "validateOnChange": true,
                          "validateOnMount": false,
                          "values": Object {
                            "artist": "",
                            "mood": "",
                            "password": "",
                            "title": "",
                          },
                        }
                      }
                      name="title"
                    />
                  </FormikConnect(ErrorMessageImpl)>
                </div>
              </RatingsFormItem>
              <RatingsFormItem
                currentValue=""
                fieldName="artist"
                fieldTitle="Artist"
                fieldType="text"
              >
                <div
                  className="formItem highlighted"
                  key="artist"
                >
                  Artist
                  <br />
                  <Field
                    autoComplete="off"
                    name="artist"
                    type="text"
                  >
                    <input
                      autoComplete="off"
                      name="artist"
                      onBlur={[Function]}
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </Field>
                  <FormikConnect(ErrorMessageImpl)
                    component={[Function]}
                    name="artist"
                  >
                    <ErrorMessageImpl
                      component={[Function]}
                      formik={
                        Object {
                          "dirty": false,
                          "errors": Object {},
                          "getFieldHelpers": [Function],
                          "getFieldMeta": [Function],
                          "getFieldProps": [Function],
                          "handleBlur": [Function],
                          "handleChange": [Function],
                          "handleReset": [Function],
                          "handleSubmit": [Function],
                          "initialErrors": Object {},
                          "initialStatus": undefined,
                          "initialTouched": Object {},
                          "initialValues": Object {
                            "artist": "",
                            "mood": "",
                            "password": "",
                            "title": "",
                          },
                          "isSubmitting": false,
                          "isValid": true,
                          "isValidating": false,
                          "registerField": [Function],
                          "resetForm": [Function],
                          "setErrors": [Function],
                          "setFieldError": [Function],
                          "setFieldTouched": [Function],
                          "setFieldValue": [Function],
                          "setFormikState": [Function],
                          "setStatus": [Function],
                          "setSubmitting": [Function],
                          "setTouched": [Function],
                          "setValues": [Function],
                          "status": undefined,
                          "submitCount": 0,
                          "submitForm": [Function],
                          "touched": Object {},
                          "unregisterField": [Function],
                          "validateField": [Function],
                          "validateForm": [Function],
                          "validateOnBlur": true,
                          "validateOnChange": true,
                          "validateOnMount": false,
                          "values": Object {
                            "artist": "",
                            "mood": "",
                            "password": "",
                            "title": "",
                          },
                        }
                      }
                      name="artist"
                    />
                  </FormikConnect(ErrorMessageImpl)>
                </div>
              </RatingsFormItem>
              <RatingsFormItem
                currentValue=""
                fieldName="mood"
                fieldTitle="Mood"
                fieldType="text"
              >
                <div
                  className="formItem highlighted"
                  key="mood"
                >
                  Mood
                  <br />
                  <Field
                    autoComplete="off"
                    name="mood"
                    type="text"
                  >
                    <input
                      autoComplete="off"
                      name="mood"
                      onBlur={[Function]}
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </Field>
                  <FormikConnect(ErrorMessageImpl)
                    component={[Function]}
                    name="mood"
                  >
                    <ErrorMessageImpl
                      component={[Function]}
                      formik={
                        Object {
                          "dirty": false,
                          "errors": Object {},
                          "getFieldHelpers": [Function],
                          "getFieldMeta": [Function],
                          "getFieldProps": [Function],
                          "handleBlur": [Function],
                          "handleChange": [Function],
                          "handleReset": [Function],
                          "handleSubmit": [Function],
                          "initialErrors": Object {},
                          "initialStatus": undefined,
                          "initialTouched": Object {},
                          "initialValues": Object {
                            "artist": "",
                            "mood": "",
                            "password": "",
                            "title": "",
                          },
                          "isSubmitting": false,
                          "isValid": true,
                          "isValidating": false,
                          "registerField": [Function],
                          "resetForm": [Function],
                          "setErrors": [Function],
                          "setFieldError": [Function],
                          "setFieldTouched": [Function],
                          "setFieldValue": [Function],
                          "setFormikState": [Function],
                          "setStatus": [Function],
                          "setSubmitting": [Function],
                          "setTouched": [Function],
                          "setValues": [Function],
                          "status": undefined,
                          "submitCount": 0,
                          "submitForm": [Function],
                          "touched": Object {},
                          "unregisterField": [Function],
                          "validateField": [Function],
                          "validateForm": [Function],
                          "validateOnBlur": true,
                          "validateOnChange": true,
                          "validateOnMount": false,
                          "values": Object {
                            "artist": "",
                            "mood": "",
                            "password": "",
                            "title": "",
                          },
                        }
                      }
                      name="mood"
                    />
                  </FormikConnect(ErrorMessageImpl)>
                </div>
              </RatingsFormItem>
              <RatingsFormItem
                currentValue=""
                fieldName="password"
                fieldTitle="Password"
                fieldType="password"
              >
                <div
                  className="formItem highlighted"
                  key="password"
                >
                  Password
                  <br />
                  <Field
                    autoComplete="current-password"
                    name="password"
                    type="password"
                  >
                    <input
                      autoComplete="current-password"
                      name="password"
                      onBlur={[Function]}
                      onChange={[Function]}
                      type="password"
                      value=""
                    />
                  </Field>
                  <FormikConnect(ErrorMessageImpl)
                    component={[Function]}
                    name="password"
                  >
                    <ErrorMessageImpl
                      component={[Function]}
                      formik={
                        Object {
                          "dirty": false,
                          "errors": Object {},
                          "getFieldHelpers": [Function],
                          "getFieldMeta": [Function],
                          "getFieldProps": [Function],
                          "handleBlur": [Function],
                          "handleChange": [Function],
                          "handleReset": [Function],
                          "handleSubmit": [Function],
                          "initialErrors": Object {},
                          "initialStatus": undefined,
                          "initialTouched": Object {},
                          "initialValues": Object {
                            "artist": "",
                            "mood": "",
                            "password": "",
                            "title": "",
                          },
                          "isSubmitting": false,
                          "isValid": true,
                          "isValidating": false,
                          "registerField": [Function],
                          "resetForm": [Function],
                          "setErrors": [Function],
                          "setFieldError": [Function],
                          "setFieldTouched": [Function],
                          "setFieldValue": [Function],
                          "setFormikState": [Function],
                          "setStatus": [Function],
                          "setSubmitting": [Function],
                          "setTouched": [Function],
                          "setValues": [Function],
                          "status": undefined,
                          "submitCount": 0,
                          "submitForm": [Function],
                          "touched": Object {},
                          "unregisterField": [Function],
                          "validateField": [Function],
                          "validateForm": [Function],
                          "validateOnBlur": true,
                          "validateOnChange": true,
                          "validateOnMount": false,
                          "values": Object {
                            "artist": "",
                            "mood": "",
                            "password": "",
                            "title": "",
                          },
                        }
                      }
                      name="password"
                    />
                  </FormikConnect(ErrorMessageImpl)>
                </div>
              </RatingsFormItem>
              <div
                className="formItem"
              >
                <button
                  disabled={false}
                  type="submit"
                >
                  Submit
                </button>
              </div>
            </form>
          </Form>
        </Formik>
      </VinylCollectionForm>
    </form>
  </Form>
</Formik>
`;
